### PR TITLE
CP-46157: Add unit test for `observed_components_of`

### PR DESCRIPTION
**Pull request by GabrielBuica**
Adds a unit test for `observed_components_of`.

Tests over a list of inputs fot `observed_components_of_inputs` while changing the value of
`Xapi_globs.observer_experimental_components` with:

`no_exp_comp`    -- empty `StringSet`;
`smapi_exp_comp` -- singleton `StringSet` that represents smapi.

### CODE REVIEW

File: ocaml/tests/test_observer.ml [hunk 1]
```diff
@@ -16,6 +16,19 @@ open Tracing
...
+let component_set_testable =
+  let pp fmt a =
+    Format.fprintf fmt "%s"
+      (a
+      |> ComponentSet.to_seq
+      |> Seq.map Xapi_observer_components.to_string
+      |> List.of_seq
+      |> String.concat ","
+      )
+  in
```
Comment (psafont) [inline ocaml/tests/test_observer.ml:30]:
```suggestion
let component_set_to_string cs =
  cs
  |> ComponentSet.to_seq
  |> Seq.map Xapi_observer_components.to_string
  |> List.of_seq
  |> String.concat ","

let component_set_testable = 
  let pp = Fmt.of_to_string component_set_to_string in
```

File: ocaml/xapi/xapi_observer_components.ml [hunk 2]
```diff
@@ -29,13 +29,13 @@ let to_string = function
       Constants.observer_component_smapi
 
 let of_string = function
-  | str when String.equal str Constants.observer_component_xapi ->
+  | str when str = Constants.observer_component_xapi ->
```
Comment (psafont) [inline ocaml/xapi/xapi_observer_components.ml:32]: `=` has compiler magic, I'd prefer if the specialized equal function was kept

Comment (GabrielBuica) [reply ocaml/xapi/xapi_observer_components.ml]: Reverted here https://github.com/xapi-project/xen-api/pull/5445/commits/55f9577f6a031a390f848af4dcfdb433932e78bf.

File: ocaml/tests/test_observer.ml (hunk 1 above)
Comment (GabrielBuica) [reply ocaml/tests/test_observer.ml]: Done here https://github.com/xapi-project/xen-api/pull/5445/commits/55f9577f6a031a390f848af4dcfdb433932e78bf.

**Review by mg12 — APPROVED** [review]

File: ocaml/tests/test_observer.ml [hunk 3]
```diff
@@ -508,6 +521,45 @@ let test_attribute_validation () =
...
+    [all_but_comp Xapi_clusterd all; [SMApi]; [Xapi; SMApi]; all]
+  in
+  let expected_outputs_smapi_exp_comp =
+    [
+      all |> all_but_comp Xapi_clusterd |> all_but_comp SMApi
+    ; []
+    ; [Xapi]
+    ; all_but_comp SMApi all
+    ]
+  in
```
Comment (psafont) [inline ocaml/tests/test_observer.ml:542]: When writing test data it's better to use tuples so the expected result and the tested data are as close as possible. On top of that, a name can be attached, which means a simple List.iter is enough to run these tests:

```
let test_data = [
    ("No experimental component is expected"
    , StringSet.empty
    , [all_but_comp Xapi_clusterd all; [SMApi]; [Xapi; SMApi]; all]
    )
  ; ("SMapi is an experimental component"
      [...]
    )
]
List.iter test_exp_comp test_data
```

Comment (GabrielBuica) [reply ocaml/tests/test_observer.ml]: Done here e8d47962c6f3917a7235ceb069a0588d515ddfeb.

**Review by psafont — COMMENTED** [review]
Sorry for dropping the ball. The general approach seems good, but I'd like some clarifications

File: ocaml/tests/test_observer.ml [hunk 4]
```diff
@@ -508,6 +521,49 @@ let test_attribute_validation () =
...
+      , [all_but_comp Xapi_clusterd all; [SMApi]; [Xapi; SMApi]; all]
+      )
+    ; ( "SMapi is experimental component"
+      , StringSet.singleton Constants.observer_component_smapi
+      , [
+          all |> all_but_comp Xapi_clusterd |> all_but_comp SMApi
+        ; []
+        ; [Xapi]
+        ; all_but_comp SMApi all
+        ]
```
[the remark this answers is not in the document]
Comment (GabrielBuica) [reply ocaml/tests/test_observer.ml]: The length of the list of inputs that I am testing is 4. The arrays contain the respective outputs of those 4 inputs.

**Comment by GabrielBuica** [comment]
@psafont Thank you for your comments! I have addressed them here https://github.com/xapi-project/xen-api/pull/5445/commits/2bbf6b18b7d731b0dfd6134ddce8e121f862495b.

**Review by psafont — APPROVED** [review]

**Pull request merged**